### PR TITLE
In-house parser

### LINKED ISSUE #16: Parse YAML config on CLI side
Right now CLI uses GRPC endpoint of Cirrus CI (https://grpc.cirrus-ci.com/) in order to parse `.cirrus.yml` config file. The backend is written in Kotlin which makes it difficult to reuse it in Go. In order to make the CLI 100% independent from Cirrus CI and have option to run it offline or in a network without public internet we should eventually implement parsing of the YAML config in Go on the CLI side.

This is a not a priority at the moment but lets put it on the table and keep it in mind.

### PR DESCRIPTION
**Pull request by edigaryev**
Some things that still clearly need to be done:

* support `only_if` (this is currently the reason why the test suite fails)
* pass environment where it's needed for expansion
* support missing fields that are still needed for the CLI, the rest will be handled via https://github.com/cirruslabs/cirrus-cli/issues/59.

Resolves #16.

### CODE REVIEW

**Comment by edigaryev** [comment]
@fkorotkov's there are a few files that are technically still under MIT since I've taken them directly from [keanu](https://github.com/edigaryev/keanu):

* https://github.com/cirruslabs/cirrus-cli/blob/49fa7d927b9df232ca8028b572e36d980b2a806a/pkg/parser/modifier/matrix/expander.go#L1
* https://github.com/cirruslabs/cirrus-cli/blob/49fa7d927b9df232ca8028b572e36d980b2a806a/pkg/parser/modifier/matrix/matrix.go#L1

Do you consent to re-license the code in these files under MPL-2.0, so I can remove these lines?

**Comment by fkorotkov** [comment]
I consent but with MIT I don't think you even need to ask 😅

File: pkg/parser/parser.go [hunk 1]
```diff
@@ -154,6 +197,138 @@ func (p *Parser) ParseFromFile(path string) (*Result, error) {
...
 	// Parse again with the file contents supplied
-	p.FilesContents = filesContents
+	p.filesContents = filesContents
 	return p.Parse(string(config))
 }
+
+func (p *Parser) ContentHash() string {
+	digest := sha256.New()
+
+	for key, value := range p.filesContents {
```
Comment (fkorotkov) [inline pkg/parser/parser.go:207]: This doesn't look right. Content hash is per file with docker arguments and not for all Dockerfiles.

File: pkg/parser/node/finder.go [hunk 2]
```diff
@@ -0,0 +1,41 @@
+package node
+
+func (node *Node) DeepFindChildren(name string) *Node {
```
[the remark this answers is not in the document]
Comment (edigaryev) [reply pkg/parser/node/finder.go]: Fixed in https://github.com/cirruslabs/cirrus-cli/pull/58/commits/b4555613467d130c3fc113e7dfb9116fa915d49b.

File: pkg/parser/node/finder.go [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+	}
+
+	if !fulfilledAtLeastOnce {
+		return nil
+	}
+
+	return &virtualNode
+}
+
+func (node *Node) HasChildren(name string) bool {
```
[the remark this answers is not in the document]
Comment (edigaryev) [reply pkg/parser/node/finder.go]: Fixed in https://github.com/cirruslabs/cirrus-cli/pull/58/commits/b4555613467d130c3fc113e7dfb9116fa915d49b.

File: pkg/parser/parseable/default.go [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+	"github.com/cirruslabs/cirrus-cli/pkg/parser/nameable"
+	"github.com/lestrrat-go/jsschema"
+)
+
+type DefaultParser struct {
+	fields            []Field
+	collectibleFields []CollectibleField
+}
+
+func (collectible *DefaultParser) OptionalField(name nameable.Nameable, schema *schema.Schema, onFound nodeFunc) {
```
[the remark this answers is not in the document]
Comment (edigaryev) [reply pkg/parser/parseable/default.go]: Indeed, see https://github.com/cirruslabs/cirrus-cli/pull/58/commits/117cc8b94b162846b4c44ba4cc8c96c2e9dd93d1.

File: pkg/parser/parser.go (hunk 1 above)
Comment (edigaryev) [reply pkg/parser/parser.go]: Good catch, fixed in https://github.com/cirruslabs/cirrus-cli/pull/58/commits/6302f7db06b8cefddafe2c7e7260253bf9d6a70a.

**Comment by edigaryev** [comment]
>support `only_if` (this is currently the reason why the test suite fails)

Implemented in https://github.com/cirruslabs/cirrus-cli/pull/58/commits/2beda3545b5a7fa2d118c8c6e67e6d4f7a421096.

However, since some changes to the [Gval](https://github.com/PaesslerAG/gval) package are still waiting to be upstreamed (https://github.com/PaesslerAG/gval/pull/38), `go.mod` temporarily points to https://github.com/edigaryev/gval/tree/expose-base-pieces.

**Comment by fkorotkov** [comment]
I think replacing the RPC parser with the local one right away seem like a bit too bold. What do you think about having a `--parser` flag with the default `remote` value and an option to set it to local? Also that way `parser_test.go` can validate that local parser returns the same results as the remote one.

**Review by fkorotkov — APPROVED** [review]

**Pull request merged**